swift/Arrow/Sources/Arrow/ArrowEncoder.swift

```diff
@@ -0,0 +1,464 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+public class ArrowEncoder: Encoder {
+    public private(set) var builders = [String: ArrowArrayHolderBuilder]()
+    private var byIndex = [String]()
+    public var codingPath: [CodingKey] = []
+    public var userInfo: [CodingUserInfoKey: Any] = [:]
+    var errorMsg: String?
+    // this is used for Dictionary types.  A dictionary type
+    // will give each key and value there own index so instead
+    // of having a 2 column RecordBatch you would have
+    // 2 * length(dictionary) column RecordBatch. Which would not
+    // be the expected output.
+    var modForIndex: Int?
+
+    public init() {}
+
+    public init(_ builders: [String: ArrowArrayHolderBuilder], byIndex: [String]) {
+        self.builders = builders
+        self.byIndex = byIndex
+    }
+
+    public static func encode<T: Encodable>(_ data: T) throws -> RecordBatch {
+        let encoder: ArrowEncoder?
```

Comment:
It seems that `encoder` must not `nil`. Can we remove this `?`?

```suggestion
        let encoder: ArrowEncoder
```

Reply:
Good idea!  The code for instantiating the encoder was similar in both the encode methods.  I have refactored that code into its own method and updated this encoder variable to be non Optional.